src/target/riscv/riscv-013.c

```diff
@@ -2471,6 +2471,33 @@ static int read_memory_progbuf(struct target *target, target_addr_t address,
 
 	memset(buffer, 0, count*size);
 
+	uint64_t mstatus = 0;
+	uint64_t mstatus_old = 0;
+	if (progbufsize >= 4) {
+		/* Read DCSR */
+		uint64_t dcsr;
+		if (register_read(target, &dcsr, GDB_REGNO_DCSR) != ERROR_OK)
+			return ERROR_FAIL;
+
+		/* Read and save MSTATUS */
+		if (register_read(target, &mstatus, GDB_REGNO_MSTATUS) != ERROR_OK)
```

Comment:
Is it necessary to read this if the DCSR_PRV check below fails?

Reply:
I guess so. If `dcsr.prv = 3` (m-mode), `mstatus.mprv` still could have been set from outside debug mode (e.g. to enable the MMU in m-mode). Thus we need it later for the final check whether the `csrr*i` instructions should be written into the program buffer.